pkg/provider/provider.go

```diff
@@ -0,0 +1,604 @@
+package provider
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/containous/maesh/pkg/k8s"
+	"github.com/containous/maesh/pkg/topology"
+	"github.com/containous/traefik/v2/pkg/config/dynamic"
+	"github.com/sirupsen/logrus"
+)
+
+// TopologyBuilder is capable of building Topologies.
+type TopologyBuilder interface {
+	Build(ignored k8s.IgnoreWrapper) (*topology.Topology, error)
+}
+
+// TCPPortFinder is capable for finding a TCP port mapping.
+type TCPPortFinder interface {
+	Find(svc k8s.ServiceWithPort) (int32, bool)
+}
+
+// When multiple Traefik Routers listen to the same entrypoint and have the same Rule, the chosen router will be the one
+// with the highest priority. There are few cases where this priority is crucial when building the dynamic configuration:
+// - When a TrafficSplit is set on a k8s service, this will create 2 Traefik Routers. One for accessing the k8s service
+//   endpoints and one for accessing the services endpoints mentioned in the TrafficSplit. They both have the same Rule
+//   but we should always prioritize the TrafficSplit. Therefore, TrafficSplit Routers should always have a higher priority.
+// - When a TrafficTarget Destination targets pods of a k8s service and a TrafficSplit is set on this service. This
+//   will create 2 Traefik Routers. One for the TrafficSplit and one for the TrafficTarget. We should always prioritize
+//   TrafficSplits Routers and so, TrafficSplit Routers should always have a higher priority than TrafficTarget Routers.
+const (
+	priorityService               = 1
+	priorityTrafficTargetDirect   = 2
+	priorityTrafficTargetIndirect = 3
+	priorityTrafficSplit          = 4
+)
+
+// Provider holds the configuration for generating dynamic configuration from a kubernetes cluster state.
+type Provider struct {
+	acl                bool
+	minHTTPPort        int32
+	maxHTTPPort        int32
+	defaultTrafficType string
+	maeshNamespace     string
+
+	topologyBuilder   TopologyBuilder
+	tcpStateTable     TCPPortFinder
+	middlewareBuilder MiddlewareBuilder
+	ignored           k8s.IgnoreWrapper
+
+	logger logrus.FieldLogger
+}
+
+// New creates a new Provider.
+func New(topologyBuilder TopologyBuilder, tcpStateTable TCPPortFinder, ignored k8s.IgnoreWrapper, minHTTPPort, maxHTTPPort int32, acl bool, defaultTrafficType, maeshNamespace string, logger logrus.FieldLogger) *Provider {
```

Comment:
At this point, WDYT of building a structure containing the dependencies of the Provider, so that we can have only one argument here and make the use of this `New` function more readable?

(I'm not saying we need variadic modifiers for building the provider, just a struct that we would build explicitly before calling new, which would make it clearer how it works, IMO)

Reply:
I kind of agree with you. Putting everything related to Maesh configuration in a proper struct would make more sense. So I would put:
- ignored k8s.IgnoreWrapper
- minHTTPPort
- maxHTTPPort int32
- acl bool
- defaultTrafficType string
- maeshNamespace string